Improve tests for Model._handle_message_event

kaustubh-nair opened the pull request:
This should hopefully be a fix for #642 #574 and #566

kaustubh-nair committed:
tests: model: Add failing test for unread count bug.

While handling the message event, message indexing and unread count updates
should happen even if user in the same narrow as that of the received
message. `model.found_newest` is True when the received message belongs
to the current narrow.

Currently, a message is added to index and unread counts are updated only
when `model.found_newest` is True. This leads to
bugs like negative unread counts and missing message information in the
index. This commit adds a failing test for fixing this bug.

kaustubh-nair committed:
model: Index message and set count irrespective of found_newest state.

found_newest is returned by the server indicating whether
the messages list includes the latest message in the narrow.
This commit ensures that message indexing and unread count updates
happen even if the received message is in a different narrow.

Tests amended.

neiljp commented on tests/model/test_model.py:
```diff
@@ -783,6 +783,36 @@ def test__handle_message_event(self, mocker, user_profile, response,
+          'id': 1}, 'flags': []}),
+        ])
+    def test__handle_message_event_set_count_all_narrows(self, mocker, event,
+                                                         model, found_newest,
```
This uses `model`, which has an initial default narrow of 'all messages', so this doesn't replicate the 'in a different narrow' test case we started with. This didn't turn out to be the issue, so I'm not sure if we want to explore behavior over different narrows.

kaustubh-nair replied:
Being in a different narrow won't make a difference since it's only `found_newest` which affects behaviour?

kaustubh-nair replied:
Renamed the test to `test__handle_message_event_set_count_found_newest`, since we are not directly considering different narrows.

PS. Now that I look at this, it feels a little too much to write such a big test for such a small change.
Wouldn't a comment saying "Update counts irrespective of found_newest" have been enough?

neiljp commented on zulipterminal/model.py:
```diff
@@ -727,8 +727,10 @@ def _handle_message_event(self, event: Event) -> None:
 
+        self.index = index_messages([message], self, self.index)
+        if 'read' not in message['flags']:
+            set_count([message['id']], self.controller, 1)
```
minor: follow this with a newline, and possibly add a comment to this block re updating counts and how the message needs to be in the index to be identified by `set_count`.

From other conversations, I wonder if we might update `set_count` to just accept the message(s) themselves to avoid an index dependence, but that's another discussion. That would avoid the coupling between these two actions (add to index, set count), as the message need not necessarily be in the index.

neiljp commented on tests/model/test_model.py:
```diff
@@ -785,8 +785,7 @@ def test__handle_message_event(self, mocker, user_profile, response,
         (True),
-        pytest.param(False,
-                     marks=pytest.mark.xfail(reason="Unread count bug")),
+        (False),
```
I understand this is a simple adjustment from the previous commit and so a small diff, but I wonder if this should shrink onto one line now.

neiljp commented on tests/model/test_model.py:
```diff
@@ -783,6 +783,36 @@ def test__handle_message_event(self, mocker, user_profile, response,
         assert model.msg_list.log == log
 
+    @pytest.mark.parametrize('found_newest', [
+        (True),
```
For a one-parameter entry, this is just `True`, ie. the parentheses are unnecessary.

neiljp reviewed (commented):
@kaustubh-nair Thanks for filing this and working on the extra test :+1: 

Some of my points will disappear if we squash the two commits.

Other points are discussion points that we could use to explore other improvements.

kaustubh-nair commented on tests/model/test_model.py:
```diff
@@ -783,6 +783,36 @@ def test__handle_message_event(self, mocker, user_profile, response,
+        index_msg.assert_called_once_with([event['message']], model,
+                                          model.index)
+        set_count.assert_called_once_with([event['message']['id']],
+                                          self.controller, 1)
```
@neiljp As in, change in state of unread counts and index?

kaustubh-nair commented:
@neiljp Thanks for the review, I've pushed the updated commits.

sumanthvrao commented:
@kaustubh-nair Thanks for working on this 👍 . I tried some manual testing and the results are promising.

One follow-up point worth mentioning here (in case you or someone else wants to pick it) is to try and bridge the 'holes' in index forming which this might trigger. For instance, the new logic  will solve the negative unread_count issues we have been seeing, but suppose we index a non-current message which is received and then the user scrolls down (keypress down) causing new message to be fetched, how will the index handle this? Will the re-index cause any problem since the message may potentially exist already in the index.

This don't need to be solved right away, but it's worth keeping in mind of the impact this might have on the indexing changes.

neiljp commented:
@kaustubh-nair I just pushed the code part of this PR, since this is an important fix that I don't want to leave unmerged for too long - both because it improves the overall experience, and since it may expose other bugs.

Let's discuss whether you want to leave this PR open to improve the tests for this method, or whether to add this as another issue so we don't forget it.

kaustubh-nair commented:
Pushing some commits for refactoring tests now.

PS. If you think the current commits don't require any more changes, you can merge them. I'll add commits on top.

kaustubh-nair committed:
tests: refactor: Create new fixture for __handle_message_event tests.

Extract common attributes and attributes in a new fixture to decrease
duplication.

kaustubh-nair committed:
tests: refactor: Remove unused variable from test.

Remove create_mst_box_list from test__handle_message_event_with_flags.

kaustubh-nair committed:
tests: refactor: Refactor test__handle_message_event_with_flags.

This test calls __handle_message_event twice, once with flags and once
without.

This commit moves the flags to a parameterize block to decrease
duplication. Additionally, it remove the assertion for
'model.notify_user()' since it is not related to this test.

kaustubh-nair committed:
tests: model: Create new test for notifying users after message event.

* Move model.notify_users() to its own test.
* Remove assertions of the method from other tests.

kaustubh-nair commented on tests/model/test_model.py:
```diff
@@ -645,17 +645,23 @@ def test__stream_info_from_subscriptions(self, initial_data, streams,
+    def test__handle_message_event_with_Falsey_log(self, mocker,
+                                                   model_message_event,
+                                                   message_fixture):
+        model = model_message_event
```
Doing this because `model_message_event` is a really long variable name.

kaustubh-nair commented:
The main commit here was merged, so renamed the PR.
Does need a rebase though.
